### PR TITLE
JDK-8187314: All Cells: must show backing data always

### PR DESCRIPTION
**Pull request by Maran23**
Before, the `updateItem` method was called with the new value that was committed via `commitEdit()`.
This is problematic as developers may setup a commit handler via `setOnEditCommit`, which may reject the edit (or change the value otherwise).
We therefore do call the `updateItem(-1)` which will also call `updateItem` but with the real underlying value.

Changed and added tests for all 4 cells.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8187314](https://bugs.openjdk.org/browse/JDK-8187314): All Cells: must show backing data always (**Bug** - P3)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1197/head:pull/1197` \
`$ git checkout pull/1197`

Update a local copy of the PR: \
`$ git checkout pull/1197` \
`$ git pull https://git.openjdk.org/jfx.git pull/1197/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1197`

View PR using the GUI difftool: \
`$ git pr show -t 1197`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1197.diff">https://git.openjdk.org/jfx/pull/1197.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1197#issuecomment-1666005627)

### CODE REVIEW

**Comment by kevinrushforth** [comment]
This seems reasonable.

@aghaisas can you review it. It will need a second reviewer as well.

/reviewers 2

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java [hunk 1]
```diff
@@ -1106,6 +1106,30 @@ public void testMisbehavingCancelEditTerminatesEdit() {
         }
     }
 
+    /**
+     * See also: <a href="https://bugs.openjdk.org/browse/JDK-8187314">JDK-8187314</a>.
+     */
+    @Test
+    public void testEditCommitValueChangeIsReflectedInCell() {
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ListCellTest.java:1113]: this test passes with the fix, fails without as expected.
but for some reason I cannot reproduce the issue with ListView in the MonkeyTester
https://github.com/andy-goryachev-oracle/MonkeyTest

(List View Page -> select 'editable' -> double click to edit, type "updated" to save changes, any other text to reject)

**Comment by andy-goryachev-oracle** [comment]
I've tested this change with ListView, TableView, and TreeView.
Do you have a SCCE for TreeTableView similar to the TreeView example in the ticket?

**Comment by andy-goryachev-oracle** [comment]
Do you have a SCCE for TreeTableView similar to the TreeView example in the ticket?

**Comment by Maran23** [comment]
> Do you have a SCCE for TreeTableView similar to the TreeView example in the ticket?

It should be the same as `TreeTableView` works basically the same as `TableView`.

**Comment by andy-goryachev-oracle** [comment]
could you please create one?

**Comment by Maran23** [comment]
As I said, everything is basically the same, but changed to `TreeItem`:

```
import javafx.application.Application;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeTableColumn;
import javafx.scene.control.TreeTablePosition;
import javafx.scene.control.TreeTableView;
import javafx.scene.control.cell.TextFieldTreeTableCell;
import javafx.scene.control.cell.TreeItemPropertyValueFactory;
import javafx.scene.layout.BorderPane;
import javafx.stage.Stage;

public class Test extends Application {

    TreeTablePosition<Dummy, String> editPosition;
    private Object editValue;
    @Override
    public void start(Stage primaryStage) {
        TreeTableView<Dummy> table = new TreeTableView<>();
        table.setRoot(new TreeItem<>());
        for (Dummy dummy : Dummy.dummies()) {
            TreeItem<Dummy> dummyItem = new TreeItem<>(dummy);

            table.getRoot().getChildren().add(dummyItem);
        }

        table.setEditable(true);

        TreeTableColumn<Dummy, String> first = new TreeTableColumn<>("Text");
        first.setCellFactory(TextFieldTreeTableCell.forTreeTableColumn());
        first.setCellValueFactory(new TreeItemPropertyValueFactory<>("dummy"));

        first.setOnEditStart(t -> editPosition = t.getTreeTablePosition());
        first.addEventHandler(TreeTableColumn.editCommitEvent(), t -> {
            editValue = t.getNewValue();
            System.out.println("doing nothing");

        });

        table.getColumns().addAll(first);

        Button button = new Button("Check value");
        button.setOnAction(e -> {
            if (editPosition == null) return;
            String value = editPosition.getTableColumn().getCellObservableValue(editPosition.getRow()).getValue();
            System.out.println(
                    "value in edited cell must represent backing data: " + value + " not the edited " + editValue);
        });
        BorderPane root = new BorderPane(table);
        root.setBottom(button);
        Scene scene = new Scene(root, 300, 250);

        primaryStage.setTitle("Hello World!");
        primaryStage.setScene(scene);
        primaryStage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }

    public static class Dummy {
        private String dummy;
        public Dummy(String dummy) {
            this.dummy = dummy;
        }

        public String getDummy() {
            return dummy;
        }

        public static ObservableList<Dummy> dummies() {
            return FXCollections.observableArrayList(
                    new Dummy("1"), new Dummy("2"), new Dummy("3")
            );
        }
    }
}
```

**Comment by andy-goryachev-oracle** [comment]
Thank you so much, @Maran23 !  It makes it so much easier to review.

Since the master progressed a bit since the last review, may I trouble you to merge in the latest master please?

**Comment by andy-goryachev-oracle** [comment]
Also, would it be possible to extend the tree table view test to include a second column with the same logic?

**Comment by Maran23** [comment]
> Also, would it be possible to extend the tree table view test to include a second column with the same logic?

Sure, but why? 
I mean, a test should usually consist of as few components as are needed for testing, since we want to test a specific problem (in isolation). 
And one column is sufficient to test the issue, more columns will just add more bloat but no new insights (and the issue is not related to a particular column anyway)

**Comment by andy-goryachev-oracle** [comment]
> Sure, but why?

The first column in tree table is different from subsequent columns.  I just want to make sure it works there as well.

**Comment by andy-goryachev-oracle** [comment]
TreeTableView with multiple columns:

https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/bugs/TreeTableView_8187314.java

works as expected.

**Review by andy-goryachev-oracle — APPROVED** [review]
🕶 looks good, thanks!

**Review by aghaisas — APPROVED** [review]

**Comment by Maran23** [comment]
> TreeTableView with multiple columns works as expected.

Thanks for checking. In the meantime I also checked out the tests, and all of them do test one column (`startEdit`, `cancelEdit`, `commitEdit`, `setOnEditCommit`, ....), so I really think we are good here! 🙂

**Pull request closed without merging**